payjoin/src/receive/mod.rs

```diff
@@ -686,13 +673,78 @@ impl ProvisionalProposal {
                     .iter()
                     .position(|txo| txo.script_pubkey == sender_fee_output.script_pubkey)
                     .expect("Sender output is missing from payjoin PSBT");
+                // Determine the additional amount that the sender will pay in fees
+                let sender_additional_fee = min(max_additional_fee_contribution, additional_fee);
+                log::trace!("sender_additional_fee: {}", sender_additional_fee);
                 // Remove additional miner fee from the sender's specified output
-                self.payjoin_psbt.unsigned_tx.output[sender_fee_vout].value -= additional_fee;
+                self.payjoin_psbt.unsigned_tx.output[sender_fee_vout].value -=
+                    sender_additional_fee;
+                receiver_additional_fee -= sender_additional_fee;
             }
+
+            // The receiver covers any additional fees if applicable
+            if receiver_additional_fee > Amount::ZERO {
+                log::trace!("receiver_additional_fee: {}", receiver_additional_fee);
+                // Remove additional miner fee from the receiver's specified output
+                self.payjoin_psbt.unsigned_tx.output[self.change_vout].value -=
+                    receiver_additional_fee;
+            }
+        }
+
+        // The sender's fee contribution can only be used to pay for additional input weight, so
+        // any additional outputs must be paid for by the receiver.
+        let additional_receiver_fee = self.additional_output_fee(min_feerate);
+        if additional_receiver_fee > Amount::ZERO {
+            // Remove additional miner fee from the receiver's specified output
+            self.payjoin_psbt.unsigned_tx.output[self.change_vout].value -= additional_receiver_fee;
         }
         Ok(&self.payjoin_psbt)
     }
 
+    /// Calculate the additional fee required to pay for any additional inputs in the payjoin tx
+    fn additional_input_fee(&self, min_feerate: FeeRate) -> Result<Amount, RequestError> {
+        // This error should never happen. We check for at least one input in the constructor
+        let input_pair = self
+            .payjoin_psbt
+            .input_pairs()
+            .next()
+            .ok_or(InternalRequestError::OriginalPsbtNotBroadcastable)?;
+        // Calculate the additional fee contribution
+        let txo = input_pair.previous_txout().map_err(InternalRequestError::PrevTxOut)?;
+        let input_type = InputType::from_spent_input(txo, &self.payjoin_psbt.inputs[0])
+            .map_err(InternalRequestError::InputType)?;
+        let input_count = self.payjoin_psbt.inputs.len() - self.original_psbt.inputs.len();
+        log::trace!("input_count : {}", input_count);
+        let weight_per_input = input_type.expected_input_weight();
```

Comment:
Do expected_input_weights need to be adjusted? Or worst case scenario the receiver overestimates by a few vB weight? e.g. [is lopp's estimate for p2wpkh output is 67.75](https://github.com/jlopp/bitcoin-transaction-size-calculator/blob/6d3143c02493fce2230f6d3903309bb1576fc079/index.html#L26) appropriate for us to use?;

Reply:
I believe `expected_input_weight()` is correct, there is a comment a few lines above in the lopp code that explains the difference: `All segwit input sizes are reduced by 1 WU to account for the witness item counts being added for every input per the transaction header`

I think we could remove all our custom weight calculation code and use weight predictions from `rust-bitcoin` as you suggested earlier, but I think that's better to tackle as a separate PR. I'll create a follow-up issue for this.